### PR TITLE
Add `uniqueString`

### PR DESCRIPTION
**Pull request by michsa**
Adds `uniqueString`, which deduplicates strings, and for good measure, `uniqueStringHash`, which manages multiple `uniqueString`s.

I'm not very familiar with writing functions in this style, so I might have gone a little overboard with it 😅

### CODE REVIEW

**Comment by decrapifier** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Messages</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:book:</td>
      <td>Your PR has no browser errors. Great job!</td>
    </tr>
  
<tr>
      <td>:book:</td>
      <td>We were able to automatically fix some formatting issues in this PR for you!</td>
    </tr>
  </tbody>
</table>


### Some things that were possibly fixed:

- Code that could be fixed via the --fix flag
- Formatting that could be fixed by prettier

Take a look at this commit to see what happened in detail: 2ab5629522c6856cfbf95738075a5a2007bc7b3d


And look at this wiki page to see the reasoning behind the ESLint rules: https://github.com/smartprocure/eslint-config-smartprocure/wiki/Rules-and-Why-We-Chose-Them
<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against dca670aff88a69b45025ecea797537841e495352
</p>

File: README.md [hunk 1]
```diff
@@ -415,6 +415,23 @@ Example: `['a', 'b', 'c'] -> 'a, b and c'`
...
 
+### uniqueString
+
+`array -> string -> string` Returns a function that takes a string and de-duplicates it against an internal cache. Each time this function is called, the resulting deduplicated string is added to the cache. Exposes `cache` and `clear()` properties to access and clear the cache, respectively.
+
+Example: `_.map(uniqueString([]), ['foo', 'foo', 'foo']) -> ['foo', 'foo1', 'foo2']`
+
+### uniqueStringWith
+
+`(cachizer, array) -> string -> string` Allows passing a function (`array -> object`) to override the way `uniqueString`'s initial array is converted into a cache object. Can be curried with a cachizer to create a custom `uniqueString` function.
```
Comment (daedalus28) [inline README.md]: `cachizer` might be better named as something indicating that it's about counting occurences 🤔

File: README.md [hunk 2]
```diff
@@ -415,6 +415,23 @@ Example: `['a', 'b', 'c'] -> 'a, b and c'`
 
 Example: `(' - ', ' or ', ['a', 'b', 'c']) -> 'a - b or c'`
 
+### uniqueString
+
+`array -> string -> string` Returns a function that takes a string and de-duplicates it against an internal cache. Each time this function is called, the resulting deduplicated string is added to the cache. Exposes `cache` and `clear()` properties to access and clear the cache, respectively.
+
+Example: `_.map(uniqueString([]), ['foo', 'foo', 'foo']) -> ['foo', 'foo1', 'foo2']`
```
Comment (daedalus28) [inline README.md]: we should add an example with clearing/accessing the cache

File: README.md (hunk 1 above)
Comment (daedalus28) [reply README.md]: we also probably should mention that it also has cache/clear and maybe give an example

Comment (michsa) [reply README.md]: It's not necessarily about counting - if there are duplicates in the input array, they only really need to be counted once for `uniqueString` to dedupe properly, so `F.arrayToObject(keyFn, () => 1)` is often just as useful (if not more so) than `_.countBy(keyFn)`. eg:

```js
let dedupeWithCount = uniqueStringWith(
  _.countBy(_.identity), ['foo', 'foo', 'foo', 'bar']
)
let dedupeNoCount = uniqueStringWith(
  F.arrayToObject(_.identity, () => 1), ['foo', 'foo', 'foo', 'bar']
)

_.map(dedupeWithCount, ['foo', 'bar'])  //-> ['foo3', 'bar1']
_.map(dedupeNoCount, ['foo', 'bar'])  //-> ['foo1', 'bar1']
```

These both work fine, but I'd argue that the second one is more intuitive ("what happened to `foo1` and `foo2`?")

So I'm gonna stand by calling it `cachizer` for now - but, yeah, some more details about the function would be a good idea anyway 🙂

**Pull request merged**